File: training-data/nim/bellman_ford.nim
type Edge = tuple[src, dst, weight: int]

proc bellmanFord(edges: seq[Edge], vertexCount, source: int): seq[int] =
  var dist = newSeq[int](vertexCount)
  for i in 0 ..< vertexCount:
    dist[i] = high(int)
  dist[source] = 0
  for i in 0 ..< vertexCount - 1:
    for e in edges:
      if dist[e.src] != high(int) and dist[e.src] + e.weight < dist[e.dst]:
        dist[e.dst] = dist[e.src] + e.weight
  for e in edges:
    if dist[e.src] != high(int) and dist[e.src] + e.weight < dist[e.dst]:
      raise newException(ValueError, "graph contains a negative-weight cycle")
  result = dist

let edges = @[
  (src: 0, dst: 1, weight: 4),
  (src: 0, dst: 2, weight: 1),
  (src: 2, dst: 1, weight: 2),
  (src: 1, dst: 3, weight: 1),
  (src: 2, dst: 3, weight: 5)
]
echo bellmanFord(edges, 4, 0)

File: training-data/perl/bellman_ford.pl
use strict;
use warnings;

sub bellman_ford {
    my ($edges, $vertices, $source) = @_;
    my %dist = map { $_ => 1e18 } @$vertices;
    $dist{$source} = 0;

    for my $i (1 .. scalar(@$vertices) - 1) {
        for my $edge (@$edges) {
            my ($u, $v, $w) = @$edge;
            if ($dist{$u} + $w < $dist{$v}) {
                $dist{$v} = $dist{$u} + $w;
            }
        }
    }

    for my $edge (@$edges) {
        my ($u, $v, $w) = @$edge;
        if ($dist{$u} + $w < $dist{$v}) {
            die "graph contains a negative-weight cycle";
        }
    }

    return \%dist;
}

my @vertices = qw(a b c d e);
my @edges = (
    ['a', 'b', -1], ['a', 'c', 4], ['b', 'c', 3],
    ['b', 'd', 2],  ['b', 'e', 2], ['d', 'b', 1],
    ['d', 'c', 5],  ['e', 'd', -3],
);

my $dist = bellman_ford(\@edges, \@vertices, 'a');
for my $v (@vertices) {
    print "$v: $dist->{$v}\n";
}
